### PR TITLE
Cache subrings returned by `Ring.ShuffleShardWithLookback()`

### PR DESCRIPTION
**Pull request by charleskorn**
**What this PR does**:

This PR adds caching for `Ring.ShuffleShardWithLookback()`. `ShuffleShardWithLookback` is a potentially expensive operation (see benchmarks in #281), and the results are cacheable.

As far as I can tell, the only system that uses `ShuffleShardWithLookback` is Mimir - I couldn't find any references in Loki, Tempo or Phlare. So I've implemented a cache that suits Mimir's access pattern, namely that we only request subrings with lookback windows later than previous lookback windows. For example, if we request a subring for `user-1` with lookback window starting at $T_1$, then subsequent requests for `user-1` will have a lookback window starting at or after $T_1$. 

`ShuffleShardWithLookback` will still return correct results if this access pattern is not followed, but the cache will not be effective.

Cached subrings are returned if no members of the subring registered between the beginning of the cached lookback window and the beginning of the requested lookback window. This is best illustrated visually:

ShuffleShardWithLookback caching explanation

Explanation:
* Imagine the first request to this ring is for lookback window A. The cache is empty, so we compute the subring $R_1$, store it in the cache and return it to the caller.
* Next, another request is made for lookback window B. No instances in $R_1$ registered between the start of window A and the start of window B, so we can return the cached $R_1$.
* Finally, a third request is made, this time for lookback window C. An instance in $R_1$ registered between the start of window A and the start of window C, so we can't reuse $R_1$ and instead must compute a new subring, $R_2$, store it in the cache and return it to the caller.

(Please check my logic carefully here: I'm not super familiar with the ring implementation and may have missed some edge cases.)

The cache implementation follows a similar pattern to the caching implemented for `Ring.ShuffleShard()`, and all cached subrings are discarded when the ring topology changes.

**Which issue(s) this PR fixes**:

(none)

**Checklist**
- [x] Tests updated
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by pstibrany** [comment]
tl;dr: I think my mental model is slightly different than what code says, but I can't see a problem with the code. I will spend more time with it.

I find the image bit confusing. Lookback period indicates that we are looking for recently-registered instances (within lookback period), however image indicates looking "forward".

Here is an example:

- Say we have some recently registered instances: Instance A registered at 10:00, B registered at 12:05.
- At 12:10, we do a shuffle shard subring request, with lookback period 2h. It means we're looking back up to 10:10 to see if there were any registered instances. This will find instance B. Instance A is outside of this window.
- Subring found in previous step is valid until 14:05 (instance B registration + 2h period), assuming no no other changes to the ring in the meantime.
- at time 14:20, we can no longer use cached version of the subring, and need to compute a new one. This will be valid "forever", because there were no instances registered recently.
- Let's say at 15:00 new instance C registers. Then we invalidate caches, and recompute subrings again.

Looking at the PR code, I don't understand following details:
- why do we need `validForLookbackFrom`?
- why doesn't `validForLookbackUntil` use lookback period?

(After manually stepping through tests, I think it's doing the right thing, it just doesn't quite match my mental model. I need more time to fully understand the code.)

I don't think we can reuse cached entries for different lookback periods.

For example in my example, let's say there are three requests for subring, all three are done at 13:00 and use the same identifier, but different lookback periods:

1. first requests uses 5m lookback period, and finds no registered instances between 12:55 and 13:00
2. second request uses 1h lookback period and finds one registered instance between 12:00 and 13:00 (instance B)
3. third uses 12h lookback period and finds instances A and B in lookback window 01:00 -- 13:00.

All these are valid replies and can be reused (first can be reused "forever", second can be reused until 13:05, third until 10:00+12h = 22:00). I think we should cache all three, and include lookback period (5m, 1h and 12h respectively) in the cache key.

**Comment by charleskorn** [comment]
>tl;dr: I think my mental model is slightly different than what code says, but I can't see a problem with the code. I will spend more time with it.

I think the difference in the way we're thinking about it is that I think about the lookback windows as being "starting at T, where T = now - lookback period L", whereas based on your explanation, you're thinking about them as "ending at now, looking back L". I think these are equivalent here, based on two conditions:

* we completely discard the cache when a new instance registers (which both the existing caching logic for `ShuffleShard` and the new caching logic for `ShuffleShardWithLookback` do)
* we assume that "now" is always after the last registered instance in the ring (and the existing shuffle sharding code does already assume this)

This might be why my diagram was confusing - it was based on how I think about it. The diagram was meant to show looking backward, but emphasised the start of the lookback window as that's the important timestamp for caching here.

> Looking at the PR code, I don't understand following details:
> 
> * why do we need `validForLookbackFrom`?

This is me trying to prevent an edge case we might not care about, and is probably best explained with an example. 

Imagine we have an instance that registered at 10:30, and we're operating with a two hour lookback window. We've previously cached the subring for a user that includes this new instance. It's now 12:29:59, and two goroutines, A and B, are both about to request the subring for the same user. Let's say the following happens:

* A captures the current time, 12:29:59, but is then immediately suspended before continuing any further
* Later, B starts, and the current time is now 12:30:01. `getCachedShuffledSubringWithLookback` returns `nil`, because we need to recompute the cached subring. The subring is computed and cached.
* A then resumes, and checks the cache. Without `validForLookbackFrom`, we'd return the new subring returned to B, which is not correct if A wants the subring for the two hours to 12:29:59.

> * why doesn't `validForLookbackUntil` use lookback period?

This is probably another symptom of how I'm thinking about the problem: `validForLookbackUntil` is the last starting timestamp of a lookback window that can used the cached subring. (eg. if it's 12:30 now and we're using a 2 hour lookback period, the starting timestamp of the lookback window is 10:30.) 

I might need better names here - how about something like `validForLookbackWindowsStartingAt` and `validUntilLookbackWindowsStartingAt`?

> I don't think we can reuse cached entries for different lookback periods.
>
> ...
>
> I think we should cache all three, and include lookback period (5m, 1h and 12h respectively) in the cache key.

At present, Mimir is the only system that uses `ShuffleShardWithLookback`, and always uses the same lookback period as far as I can tell, so I didn't think this was necessary. But I can add it if you'd like.

With the code as it stands, I believe it'll behave correctly with the example you gave, but the cache won't be particularly effective (it'll cache the first subring, and recompute but not store the second and third).

**Review by pracucci — APPROVED** [review]
Good job! The logic LGTM.

File: ring/ring.go [hunk 1]
```diff
@@ -877,6 +902,85 @@ func (r *Ring) setCachedShuffledSubring(identifier string, size int, subring *Ri
...
+	if !ok {
+		return nil
+	}
+
+	lookbackStart := now.Add(-lookbackPeriod).Unix()
+	if lookbackStart < cached.validForLookbackWindowsStartingAt || lookbackStart > cached.validUntilLookbackWindowsStartingAt {
+		return nil
+	}
+
+	cachedSubring := cached.subring
```
[the remark this answers is not in the document]
Comment (charleskorn) [reply ring/ring.go:921]: Resolved in e4bd156.

File: ring/ring.go [hunk 2]
```diff
@@ -877,6 +902,85 @@ func (r *Ring) setCachedShuffledSubring(identifier string, size int, subring *Ri
...
+			validForLookbackUntil = instance.RegisteredTimestamp
+		}
+	}
+
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+
+	// Only cache if *this* ring hasn't changed since computing result
+	// (which can happen between releasing the read lock and getting read-write lock).
+	// Note that shuffledSubringCache can be only nil when set by test.
```
[the remark this answers is not in the document]
Comment (charleskorn) [reply ring/ring.go]: Resolved in e4bd156.

File: ring/ring.go [hunk 3]
```diff
@@ -877,6 +902,85 @@ func (r *Ring) setCachedShuffledSubring(identifier string, size int, subring *Ri
...
+		return nil
+	}
+
+	cached, ok := r.shuffledSubringWithLookbackCache[subringCacheKey{identifier: identifier, shardSize: size, lookbackPeriod: lookbackPeriod}]
+	if !ok {
+		return nil
+	}
+
+	lookbackStart := now.Add(-lookbackPeriod).Unix()
+	if lookbackStart < cached.validForLookbackWindowsStartingAt || lookbackStart > cached.validUntilLookbackWindowsStartingAt {
```
[the remark this answers is not in the document]
Comment (charleskorn) [reply ring/ring.go]: Resolved in e4bd156.

File: ring/ring.go [hunk 4]
```diff
@@ -194,8 +195,15 @@ type Ring struct {
...
-	shardSize  int
+	identifier     string
+	shardSize      int
+	lookbackPeriod time.Duration
+}
+
+type cachedSubringWithLookback struct {
+	subring                             *Ring
+	validForLookbackWindowsStartingAt   int64 // if the lookback window is from T to S, validForLookbackWindowsStartingAt is the earliest value of T this cache entry is valid for
+	validUntilLookbackWindowsStartingAt int64 // if the lookback window is from T to S, validUntilLookbackWindowsStartingAt is the latest value of T this cache entry is valid for
```
[the remark this answers is not in the document]
Comment (charleskorn) [reply ring/ring.go]: Makes sense to me, I've changed this in e4bd156.

File: ring/ring.go [hunk 5]
```diff
@@ -877,6 +902,85 @@ func (r *Ring) setCachedShuffledSubring(identifier string, size int, subring *Ri
...
+	if r.cfg.SubringCacheDisabled {
+		return nil
+	}
+
+	cached, ok := r.shuffledSubringWithLookbackCache[subringCacheKey{identifier: identifier, shardSize: size, lookbackPeriod: lookbackPeriod}]
+	if !ok {
+		return nil
+	}
+
+	lookbackStart := now.Add(-lookbackPeriod).Unix()
```
[the remark this answers is not in the document]
Comment (charleskorn) [reply ring/ring.go]: Makes sense to me, I've changed this in e4bd156.

File: ring/ring_test.go [hunk 6]
```diff
@@ -1968,6 +1968,411 @@ func TestRing_ShuffleShardWithLookback_CorrectnessWithFuzzy(t *testing.T) {
...
+				generateRingInstanceWithInfo("instance-1", "zone-a", []uint32{userToken(userID, "zone-a", 0) + 1}, now.Add(-2*time.Hour)),
+				generateRingInstanceWithInfo("instance-2", "zone-a", []uint32{userToken(userID, "zone-a", 1) + 1}, now.Add(-2*time.Hour)),
+				generateRingInstanceWithInfo("instance-3", "zone-b", []uint32{userToken(userID, "zone-b", 0) + 1}, now.Add(-2*time.Hour)),
+				generateRingInstanceWithInfo("instance-4", "zone-b", []uint32{userToken(userID, "zone-b", 1) + 1}, now.Add(-2*time.Hour)),
+				generateRingInstanceWithInfo("instance-5", "zone-c", []uint32{userToken(userID, "zone-c", 0) + 1}, now.Add(-2*time.Hour)),
+				generateRingInstanceWithInfo("instance-6", "zone-c", []uint32{userToken(userID, "zone-c", 1) + 1}, now.Add(-2*time.Hour)),
+			},
+			test: func(t *testing.T, ring *Ring) {
+				first := ring.ShuffleShardWithLookback(userID, subringSize, time.Hour, now)
+				second := ring.ShuffleShardWithLookback(userID, subringSize*2, time.Hour, now)
```
[the remark this answers is not in the document]
Comment (charleskorn) [reply ring/ring_test.go]: Good catch, I've fixed this in d87c519.
